p2p/protocols/protocol.go

```diff
@@ -225,31 +229,103 @@ func NewPeer(peer *p2p.Peer, rw p2p.MsgReadWriter, spec *Spec) *Peer {
 	}
 }
 
-// Run starts the forever loop that handles incoming messages
-// called within the p2p.Protocol#Run function
-// the handler argument is a function which is called for each message received
+// Run starts the forever loop that handles incoming messages.
+// The handler argument is a function which is called for each message received
 // from the remote peer, a returned error causes the loop to exit
-// resulting in disconnection
+// resulting in disconnection of the protocol
 func (p *Peer) Run(handler func(ctx context.Context, msg interface{}) error) error {
-	for {
-		if err := p.handleIncoming(handler); err != nil {
-			if err != io.EOF {
-				metrics.GetOrRegisterCounter("peer.handleincoming.error", nil).Inc(1)
-				log.Error("peer.handleIncoming", "err", err)
+	wait := p.run(handler)
+	return wait()
+}
+
+// run receives messages from the peer and dispatches async routines to handle the messages
+func (p *Peer) run(handler func(ctx context.Context, msg interface{}) error) func() error {
+	p.mtx.Lock()
+	p.running = true
+	p.mtx.Unlock()
+
+	errc := make(chan error)
+	go func() {
+		for {
+			msg, err := p.readMsg()
+			if err != nil {
+				select {
+				case errc <- err:
+				default:
+				}
+
+				return
 			}
-			return err
+
+			p.mtx.RLock()
+			// if loop has been stopped, we don't dispatch any more async routines and discard (consume) the message
+			if !p.running {
+				_ = msg.Discard()
+				p.mtx.RUnlock()
+				continue
+			}
+
+			p.mtx.RUnlock()
+			p.wg.Add(1)
+			go func() {
+				defer p.wg.Done()
+				err := p.handleMsg(msg, handler)
+				if err != nil {
+					select {
+					case errc <- err:
+					default:
+					}
+				}
+			}()
 		}
+	}()
+
+	return func() error {
+		return <-errc
```

Comment:
Hmmm, the only place where this function is called is in the `run` method, which is waiting for this channel in order to return. How can it leak? I am not sure wdym 🙈 Maybe it could leak if there was multiple callers waiting, but this is not the use-case for now.

Reply:
well, if you'll look again you'll see that `errc` is only written to in case of an error, but the channel is never closed otherwise. for cleanliness' sake i'd be happy if there is a `defer close(errc)` in the main goroutine that is on line 249

Reply:
You are right, I will add it! Thanks!